File: collector-old/rate-meter.mjs
// Publishes the smoothed consumption rate (tokens/hour) per account.
//
// WHY THIS EXISTS
// The rate panel used to compute `sum_over_time(tokens[15m]) * 4` directly in
// LogQL. That is a boxcar: a burst enters the window and the value jumps, then it
// sits flat until the burst slides out and the value falls off a cliff. Measured
// on real data, the curve dropped 30x in a single 5-minute step when a session
// stopped, and peaks towered ~10x over the P75 cutline — so the line sat squashed
// at the floor of the chart and crossing it meant nothing.
//
// What the panel is for is reading speed: "right now I am fast" or "right now I am
// slow", so a human can judge whether that speed is warranted (a session firing
// many tools, or an agent that quietly spawned 30 subagents). That needs inertia:
// stopping should decelerate the curve, not teleport it to zero.
//
// So the rate is an exponentially weighted moving average. LogQL has no EWMA,
// which is why it is computed here and published back into Loki.
//
// The half-life is part of the STREAM LABELS on purpose. Loki cannot replace
// derived data, so changing the half-life would otherwise mix two different maths
// in one series; with it as a label, a new value simply starts a new stream and
// the old one ages out with retention.

const STREAM = 'claude-code-rate';
const BUCKET_S = 300;
// Same definition of "new tokens" used by the gauges and the limit references.
const TOKEN_FIELDS = ['input_tokens', 'output_tokens', 'cache_creation_tokens'];

function apiSelector(emailPattern) {
  return '{service_name="claude-code"} | event_name = `api_request` '
    + `| user_email =~ \`${emailPattern}\``;
}

function escapeRegex(value) {
  return value.replace(/[.+*?()|[\]{}\\^$`]/g, '\\$&');
}

async function lokiGet(lokiUrl, path, params) {
  const url = new URL(path, lokiUrl);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, String(value));
  const response = await fetch(url);
  if (!response.ok) throw new Error(`Loki ${response.status}: ${(await response.text()).slice(0, 160)}`);
  const body = await response.json();
  if (body.status !== 'success') throw new Error(`Loki status=${body.status}`);
  return body.data?.result ?? [];
}

// Buckets of BUCKET_S seconds, as tokens/hour. `or vector(0)` on each term is
// required twice over: it fills idle buckets with a real zero (so the EWMA decays
// instead of skipping), and it keeps one field with no samples from emptying the
// whole sum, which is how LogQL treats A + B + C.
async function rateBuckets(lokiUrl, emailPattern, fields, startS, endS) {
  const selector = apiSelector(emailPattern);
  const expr = fields
    .map((field) => `(sum(sum_over_time(${selector} | unwrap ${field} [${BUCKET_S}s])) or vector(0))`)
    .join(' + ');
  const result = await lokiGet(lokiUrl, '/loki/api/v1/query_range', {
    query: expr, start: startS, end: endS, step: BUCKET_S,
  });
  const perHour = 3600 / BUCKET_S;
  return (result[0]?.values ?? []).map(([ts, value]) => [
    Math.round(Number(ts)), Number(value) * perHour,
  ]);
}

// Loki caps a range query's point count, so long backfills are fetched a day at a
// time and stitched back together in order.
async function rateBucketsOver(lokiUrl, emailPattern, fields, fromS, toS) {
  const out = [];
  const DAY = 24 * 3600;
  for (let start = fromS; start < toS; start += DAY) {
    out.push(...await rateBuckets(lokiUrl, emailPattern, fields, start, Math.min(start + DAY, toS)));
  }
  return out;
}

function ewma(points, halfLifeS) {
  const alpha = 1 - Math.exp((-Math.LN2 * BUCKET_S) / halfLifeS);
  let value = 0;
  return points.map(([ts, raw]) => {
    value = alpha * raw + (1 - alpha) * value;
    return [ts, value];
  });
}

// Published points are aligned to the bucket grid and never re-published: the
// state remembers the last timestamp written per account, so a restart does not
// duplicate a series that Loki cannot deduplicate on its own.
export async function publishRate({ lokiUrl, halfLifeS, halfLifeLabel, backfillDays, state, log, dryRun = false }) {
  const nowS = Math.floor(Date.now() / 1000);
  const label = halfLifeLabel;
  const accounts = await lokiGet(lokiUrl, '/loki/api/v1/query', {
    query: 'sum by (user_email) (count_over_time({service_name="claude-code"} '
      + '| event_name = `api_request` [7d]))',
    time: nowS,
  });
  const emails = accounts.map((series) => series.metric?.user_email).filter(Boolean);
  if (!emails.length) return 0;

  state.ratePublished ??= {};
  const streams = [];

  for (const email of emails) {
    const key = `${label}:${email}`;
    const last = state.ratePublished[key] ?? 0;
    // Warm-up matters: an EWMA started cold reads far too low for its first
    // half-lives. Always recompute from well before the first point to publish.
    const warmup = Math.max(6 * 3600, halfLifeS * 8);
    const from = last ? last - warmup : nowS - backfillDays * 24 * 3600;
    const pattern = escapeRegex(email);
    let total; let input; let output;
    try {
      [total, input, output] = await Promise.all([
        rateBucketsOver(lokiUrl, pattern, TOKEN_FIELDS, from, nowS),
        rateBucketsOver(lokiUrl, pattern, ['input_tokens'], from, nowS),
        rateBucketsOver(lokiUrl, pattern, ['output_tokens'], from, nowS),
      ]);
    } catch (error) {
      // Isolated per account, for the same reason as the usage meter: one failing
      // account must not keep every other account's rate from being published.
      log(`rate for ${email} not measured this pass: ${error.message}`);
      continue;
    }
    const smoothTotal = ewma(total, halfLifeS);
    const byTs = {
      input: new Map(ewma(input, halfLifeS)),
      output: new Map(ewma(output, halfLifeS)),
    };
    const fresh = smoothTotal.filter(([ts]) => ts > last);
    if (!fresh.length) continue;

    streams.push({
      stream: { service_name: STREAM, user_email: email, halflife: label },
      values: fresh.map(([ts, value]) => [`${ts}000000000`, 'rate', {
        rate: String(Math.round(value)),
        rate_input: String(Math.round(byTs.input.get(ts) ?? 0)),
        rate_output: String(Math.round(byTs.output.get(ts) ?? 0)),
      }]),
    });
    if (!dryRun) state.ratePublished[key] = fresh[fresh.length - 1][0];
    if (!last) log(`rate: backfilled ${fresh.length} point(s) for ${email}`);
  }

  if (!streams.length) return 0;
  if (dryRun) {
    const n = streams.reduce((sum, s) => sum + s.values.length, 0);
    log(`--dry-run: ${n} rate point(s) NOT published`);
    return n;
  }
  const response = await fetch(new URL('/loki/api/v1/push', lokiUrl), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ streams }),
  });
  if (!response.ok) {
    log(`rate meter did not publish: ${response.status} ${(await response.text()).slice(0, 160)}`);
    return 0;
  }
  return streams.reduce((sum, s) => sum + s.values.length, 0);
}

File: src/dash-generator/internal/ratemeter/ratemeter.go
// Package ratemeter ports rate-meter.mjs: an exponentially-weighted moving
// average of token consumption per account, published back into Loki so
// the rate panel has inertia instead of a boxcar's cliff-edge jumps.
package ratemeter

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"claude-observability-dash-generator/internal/lokiclient"
	"claude-observability-dash-generator/internal/state"
)

const (
	Stream  = "claude-code-rate"
	BucketS = 300
)

var TokenFields = []string{"input_tokens", "output_tokens", "cache_creation_tokens"}

func escapeRegex(value string) string {
	var b strings.Builder
	for _, r := range value {
		switch r {
		case '.', '+', '*', '?', '(', ')', '|', '[', ']', '{', '}', '\\', '^', '$', '`':
			b.WriteByte('\\')
		}
		b.WriteRune(r)
	}
	return b.String()
}

func apiSelector(emailPattern string) string {
	return fmt.Sprintf("{service_name=\"claude-code\"} | event_name = `api_request` | user_email =~ `%s`", emailPattern)
}

// Point is one (unix-seconds timestamp, value) sample.
type Point struct {
	TsS   int64
	Value float64
}

func rateBuckets(lokiURL, emailPattern string, fields []string, startS, endS int64) ([]Point, error) {
	selector := apiSelector(emailPattern)
	parts := make([]string, len(fields))
	for i, f := range fields {
		parts[i] = fmt.Sprintf("(sum(sum_over_time(%s | unwrap %s [%ds])) or vector(0))", selector, f, BucketS)
	}
	expr := strings.Join(parts, " + ")

	series, err := lokiclient.QueryRange(lokiURL, expr, startS, endS, BucketS)
	if err != nil {
		return nil, err
	}
	if len(series) == 0 {
		return nil, nil
	}
	perHour := 3600.0 / BucketS
	out := make([]Point, 0, len(series[0].Values))
	for _, v := range series[0].Values {
		ts, err := strconv.ParseFloat(v[0], 64)
		if err != nil {
			continue
		}
		val, err := strconv.ParseFloat(v[1], 64)
		if err != nil {
			continue
		}
		out = append(out, Point{TsS: int64(math.Round(ts)), Value: val * perHour})
	}
	return out, nil
}

// Loki caps a range query's point count, so a long backfill is fetched a
// day at a time and stitched back together in order.
func rateBucketsOver(lokiURL, emailPattern string, fields []string, fromS, toS int64) ([]Point, error) {
	var out []Point
	const day = 24 * 3600
	for start := fromS; start < toS; start += day {
		end := start + day
		if end > toS {
			end = toS
		}
		pts, err := rateBuckets(lokiURL, emailPattern, fields, start, end)
		if err != nil {
			return nil, err
		}
		out = append(out, pts...)
	}
	return out, nil
}

func ewma(points []Point, halfLifeS int64) []Point {
	alpha := 1 - math.Exp(-math.Ln2*BucketS/float64(halfLifeS))
	value := 0.0
	out := make([]Point, len(points))
	for i, p := range points {
		value = alpha*p.Value + (1-alpha)*value
		out[i] = Point{TsS: p.TsS, Value: value}
	}
	return out
}

// PublishConfig configures one PublishRate call.
type PublishConfig struct {
	LokiURL       string
	HalfLifeS     int64
	HalfLifeLabel string // e.g. "20m" — goes into the stream label
	BackfillDays  int
	DryRun        bool
}

// PublishRate computes and pushes the smoothed rate for every account with
// recent activity, using and updating st.RatePublished for dedup. Returns
// the number of points published (or, under DryRun, that would have been).
func PublishRate(cfg PublishConfig, st *state.State, log func(string, ...any)) (int, error) {
	nowS := time.Now().Unix()
	accounts, err := lokiclient.Query(cfg.LokiURL,
		"sum by (user_email) (count_over_time({service_name=\"claude-code\"} | event_name = `api_request` [7d]))",
		nowS)
	if err != nil {
		return 0, err
	}
	var emails []string
	for _, s := range accounts {
		if e := s.Metric["user_email"]; e != "" {
			emails = append(emails, e)
		}
	}
	if len(emails) == 0 {
		return 0, nil
	}

	if st.RatePublished == nil {
		st.RatePublished = map[string]int64{}
	}
	var streams []lokiclient.Stream
	total := 0

	for _, email := range emails {
		key := cfg.HalfLifeLabel + ":" + email
		last := st.RatePublished[key]
		warmup := int64(6 * 3600)
		if cfg.HalfLifeS*8 > warmup {
			warmup = cfg.HalfLifeS * 8
		}
		var from int64
		if last != 0 {
			from = last - warmup
		} else {
			from = nowS - int64(cfg.BackfillDays)*24*3600
		}
		pattern := escapeRegex(email)

		totalPts, err := rateBucketsOver(cfg.LokiURL, pattern, TokenFields, from, nowS)
		if err != nil {
			log("rate for %s not measured this pass: %v", email, err)
			continue
		}
		inputPts, err := rateBucketsOver(cfg.LokiURL, pattern, []string{"input_tokens"}, from, nowS)
		if err != nil {
			log("rate for %s not measured this pass: %v", email, err)
			continue
		}
		outputPts, err := rateBucketsOver(cfg.LokiURL, pattern, []string{"output_tokens"}, from, nowS)
		if err != nil {
			log("rate for %s not measured this pass: %v", email, err)
			continue
		}

		smoothTotal := ewma(totalPts, cfg.HalfLifeS)
		inputByTs := map[int64]float64{}
		for _, p := range ewma(inputPts, cfg.HalfLifeS) {
			inputByTs[p.TsS] = p.Value
		}
		outputByTs := map[int64]float64{}
		for _, p := range ewma(outputPts, cfg.HalfLifeS) {
			outputByTs[p.TsS] = p.Value
		}

		var fresh []Point
		for _, p := range smoothTotal {
			if p.TsS > last {
				fresh = append(fresh, p)
			}
		}
		if len(fresh) == 0 {
			continue
		}

		values := make([]lokiclient.StreamValue, 0, len(fresh))
		for _, p := range fresh {
			values = append(values, lokiclient.StreamValue{
				TimestampNs: fmt.Sprintf("%d000000000", p.TsS),
				Line:        "rate",
				Metadata: map[string]string{
					"rate":        strconv.Itoa(int(math.Round(p.Value))),
					"rate_input":  strconv.Itoa(int(math.Round(inputByTs[p.TsS]))),
					"rate_output": strconv.Itoa(int(math.Round(outputByTs[p.TsS]))),
				},
			})
		}
		streams = append(streams, lokiclient.Stream{
			Labels: map[string]string{"service_name": Stream, "user_email": email, "halflife": cfg.HalfLifeLabel},
			Values: values,
		})
		if !cfg.DryRun {
			st.RatePublished[key] = fresh[len(fresh)-1].TsS
		}
		if last == 0 {
			log("rate: backfilled %d point(s) for %s", len(fresh), email)
		}
		total += len(fresh)
	}

	if len(streams) == 0 {
		return 0, nil
	}
	if cfg.DryRun {
		log("--dry-run: %d rate point(s) NOT published", total)
		return total, nil
	}
	if err := lokiclient.Push(cfg.LokiURL, streams); err != nil {
		log("rate meter did not publish: %v", err)
		return 0, nil
	}
	return total, nil
}
